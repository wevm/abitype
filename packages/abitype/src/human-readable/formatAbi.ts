import type { Abi } from '../abi.js'
import { type FormatAbiItem, formatAbiItem } from './formatAbiItem.js'

/**
 * Parses JSON ABI into human-readable ABI
 *
 * @param abi - ABI
 * @returns Human-readable ABI
 */
export type FormatAbi<abi extends Abi | readonly unknown[]> = Abi extends abi
  ? readonly string[]
  : abi extends readonly []
    ? never
    : abi extends Abi
      ? {
          [key in keyof abi]: FormatAbiItem<abi[key]>
        }
      : readonly string[]

/**
 * Parses JSON ABI into human-readable ABI
 *
 * @param abi - ABI
 * @returns Human-readable ABI
 */
export function formatAbi<const abi extends Abi | readonly unknown[]>(
  abi: abi,
): FormatAbi<abi> {
  const signatures = []
  const length = abi.length
  for (let i = 0; i < length; i++) {
    const abiItem = abi[i]!
    const signature = formatAbiItem(abiItem as Abi[number])
    signatures.push(signature)
  }
  return signatures as unknown as FormatAbi<abi>
}
