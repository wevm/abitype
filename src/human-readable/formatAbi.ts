import type { Abi } from '../abi.js'
import { type FormatAbiItem, formatAbiItem } from './formatAbiItem.js'

/**
 * Parses JSON ABI into human-readable ABI
 *
 * @param TAbi - ABI
 * @returns Human-readable ABI
 */
export type FormatAbi<TAbi extends Abi | readonly unknown[]> = Abi extends TAbi
  ? readonly string[]
  : TAbi extends readonly []
  ? never
  : TAbi extends Abi
  ? {
      [K in keyof TAbi]: FormatAbiItem<TAbi[K]>
    }
  : readonly string[]

/**
 * Parses JSON ABI into human-readable ABI
 *
 * @param abi - ABI
 * @returns Human-readable ABI
 */
export function formatAbi<const TAbi extends Abi | readonly unknown[]>(
  abi: TAbi,
): FormatAbi<TAbi> {
  const signatures = []
  const length = abi.length
  for (let i = 0; i < length; i++) {
    const abiItem = abi[i]!
    const signature = formatAbiItem(abiItem as Abi[number])
    signatures.push(signature)
  }
  return signatures as unknown as FormatAbi<TAbi>
}
