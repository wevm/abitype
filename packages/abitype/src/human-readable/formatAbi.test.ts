import { expect, test } from 'vitest'

import { customSolidityErrorsAbi, seaportAbi } from '../abis/json.js'
import { formatAbi } from './formatAbi.js'

test('formatAbi', () => {
  const result = formatAbi(seaportAbi)
  expect(result).toMatchSnapshot()

  expect(formatAbi(customSolidityErrorsAbi)).toMatchInlineSnapshot(`
    [
      "constructor()",
      "error ApprovalCallerNotOwnerNorApproved()",
      "error ApprovalQueryForNonexistentToken()",
    ]
  `)
})
