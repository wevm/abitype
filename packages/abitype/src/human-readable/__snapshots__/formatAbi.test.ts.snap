// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`formatAbi 1`] = `
[
  "constructor(address conduitController)",
  "function cancel((address offerer, address zone, (uint8 itemType, address token, uint256 identifierOrCriteria, uint256 startAmount, uint256 endAmount)[] offer, (uint8 itemType, address token, uint256 identifierOrCriteria, uint256 startAmount, uint256 endAmount, address recipient)[] consideration, uint8 orderType, uint256 startTime, uint256 endTime, bytes32 zoneHash, uint256 salt, bytes32 conduitKey, uint256 counter)[] orders) returns (bool cancelled)",
  "function fulfillAdvancedOrder(((address offerer, address zone, (uint8 itemType, address token, uint256 identifierOrCriteria, uint256 startAmount, uint256 endAmount)[] offer, (uint8 itemType, address token, uint256 identifierOrCriteria, uint256 startAmount, uint256 endAmount, address recipient)[] consideration, uint8 orderType, uint256 startTime, uint256 endTime, bytes32 zoneHash, uint256 salt, bytes32 conduitKey, uint256 totalOriginalConsiderationItems) parameters, uint120 numerator, uint120 denominator, bytes signature, bytes extraData) advancedOrder, (uint256 orderIndex, uint8 side, uint256 index, uint256 identifier, bytes32[] criteriaProof)[] criteriaResolvers, bytes32 fulfillerConduitKey, address recipient) payable returns (bool fulfilled)",
  "function fulfillAvailableAdvancedOrders(((address offerer, address zone, (uint8 itemType, address token, uint256 identifierOrCriteria, uint256 startAmount, uint256 endAmount)[] offer, (uint8 itemType, address token, uint256 identifierOrCriteria, uint256 startAmount, uint256 endAmount, address recipient)[] consideration, uint8 orderType, uint256 startTime, uint256 endTime, bytes32 zoneHash, uint256 salt, bytes32 conduitKey, uint256 totalOriginalConsiderationItems) parameters, uint120 numerator, uint120 denominator, bytes signature, bytes extraData)[] advancedOrders, (uint256 orderIndex, uint8 side, uint256 index, uint256 identifier, bytes32[] criteriaProof)[] criteriaResolvers, (uint256 orderIndex, uint256 itemIndex)[][] offerFulfillments, (uint256 orderIndex, uint256 itemIndex)[][] considerationFulfillments, bytes32 fulfillerConduitKey, address recipient, uint256 maximumFulfilled) payable returns (bool[] availableOrders, ((uint8 itemType, address token, uint256 identifier, uint256 amount, address recipient) item, address offerer, bytes32 conduitKey)[] executions)",
  "function fulfillAvailableOrders(((address offerer, address zone, (uint8 itemType, address token, uint256 identifierOrCriteria, uint256 startAmount, uint256 endAmount)[] offer, (uint8 itemType, address token, uint256 identifierOrCriteria, uint256 startAmount, uint256 endAmount, address recipient)[] consideration, uint8 orderType, uint256 startTime, uint256 endTime, bytes32 zoneHash, uint256 salt, bytes32 conduitKey, uint256 totalOriginalConsiderationItems) parameters, bytes signature)[] orders, (uint256 orderIndex, uint256 itemIndex)[][] offerFulfillments, (uint256 orderIndex, uint256 itemIndex)[][] considerationFulfillments, bytes32 fulfillerConduitKey, uint256 maximumFulfilled) payable returns (bool[] availableOrders, ((uint8 itemType, address token, uint256 identifier, uint256 amount, address recipient) item, address offerer, bytes32 conduitKey)[] executions)",
  "function fulfillBasicOrder((address considerationToken, uint256 considerationIdentifier, uint256 considerationAmount, address offerer, address zone, address offerToken, uint256 offerIdentifier, uint256 offerAmount, uint8 basicOrderType, uint256 startTime, uint256 endTime, bytes32 zoneHash, uint256 salt, bytes32 offererConduitKey, bytes32 fulfillerConduitKey, uint256 totalOriginalAdditionalRecipients, (uint256 amount, address recipient)[] additionalRecipients, bytes signature) parameters) payable returns (bool fulfilled)",
  "function fulfillBasicOrder_efficient_6GL6yc((address considerationToken, uint256 considerationIdentifier, uint256 considerationAmount, address offerer, address zone, address offerToken, uint256 offerIdentifier, uint256 offerAmount, uint8 basicOrderType, uint256 startTime, uint256 endTime, bytes32 zoneHash, uint256 salt, bytes32 offererConduitKey, bytes32 fulfillerConduitKey, uint256 totalOriginalAdditionalRecipients, (uint256 amount, address recipient)[] additionalRecipients, bytes signature) parameters) payable returns (bool fulfilled)",
  "function fulfillOrder(((address offerer, address zone, (uint8 itemType, address token, uint256 identifierOrCriteria, uint256 startAmount, uint256 endAmount)[] offer, (uint8 itemType, address token, uint256 identifierOrCriteria, uint256 startAmount, uint256 endAmount, address recipient)[] consideration, uint8 orderType, uint256 startTime, uint256 endTime, bytes32 zoneHash, uint256 salt, bytes32 conduitKey, uint256 totalOriginalConsiderationItems) parameters, bytes signature) order, bytes32 fulfillerConduitKey) payable returns (bool fulfilled)",
  "function getContractOffererNonce(address contractOfferer) view returns (uint256 nonce)",
  "function getCounter(address offerer) view returns (uint256 counter)",
  "function getOrderHash((address offerer, address zone, (uint8 itemType, address token, uint256 identifierOrCriteria, uint256 startAmount, uint256 endAmount)[] offer, (uint8 itemType, address token, uint256 identifierOrCriteria, uint256 startAmount, uint256 endAmount, address recipient)[] consideration, uint8 orderType, uint256 startTime, uint256 endTime, bytes32 zoneHash, uint256 salt, bytes32 conduitKey, uint256 counter) order) view returns (bytes32 orderHash)",
  "function getOrderStatus(bytes32 orderHash) view returns (bool isValidated, bool isCancelled, uint256 totalFilled, uint256 totalSize)",
  "function incrementCounter() returns (uint256 newCounter)",
  "function information() view returns (string version, bytes32 domainSeparator, address conduitController)",
  "function matchAdvancedOrders(((address offerer, address zone, (uint8 itemType, address token, uint256 identifierOrCriteria, uint256 startAmount, uint256 endAmount)[] offer, (uint8 itemType, address token, uint256 identifierOrCriteria, uint256 startAmount, uint256 endAmount, address recipient)[] consideration, uint8 orderType, uint256 startTime, uint256 endTime, bytes32 zoneHash, uint256 salt, bytes32 conduitKey, uint256 totalOriginalConsiderationItems) parameters, uint120 numerator, uint120 denominator, bytes signature, bytes extraData)[] orders, (uint256 orderIndex, uint8 side, uint256 index, uint256 identifier, bytes32[] criteriaProof)[] criteriaResolvers, ((uint256 orderIndex, uint256 itemIndex)[] offerComponents, (uint256 orderIndex, uint256 itemIndex)[] considerationComponents)[] fulfillments, address recipient) payable returns (((uint8 itemType, address token, uint256 identifier, uint256 amount, address recipient) item, address offerer, bytes32 conduitKey)[] executions)",
  "function matchOrders(((address offerer, address zone, (uint8 itemType, address token, uint256 identifierOrCriteria, uint256 startAmount, uint256 endAmount)[] offer, (uint8 itemType, address token, uint256 identifierOrCriteria, uint256 startAmount, uint256 endAmount, address recipient)[] consideration, uint8 orderType, uint256 startTime, uint256 endTime, bytes32 zoneHash, uint256 salt, bytes32 conduitKey, uint256 totalOriginalConsiderationItems) parameters, bytes signature)[] orders, ((uint256 orderIndex, uint256 itemIndex)[] offerComponents, (uint256 orderIndex, uint256 itemIndex)[] considerationComponents)[] fulfillments) payable returns (((uint8 itemType, address token, uint256 identifier, uint256 amount, address recipient) item, address offerer, bytes32 conduitKey)[] executions)",
  "function name() view returns (string contractName)",
  "function validate(((address offerer, address zone, (uint8 itemType, address token, uint256 identifierOrCriteria, uint256 startAmount, uint256 endAmount)[] offer, (uint8 itemType, address token, uint256 identifierOrCriteria, uint256 startAmount, uint256 endAmount, address recipient)[] consideration, uint8 orderType, uint256 startTime, uint256 endTime, bytes32 zoneHash, uint256 salt, bytes32 conduitKey, uint256 totalOriginalConsiderationItems) parameters, bytes signature)[] orders) returns (bool validated)",
  "error BadContractSignature()",
  "error BadFraction()",
  "error BadReturnValueFromERC20OnTransfer(address token, address from, address to, uint256 amount)",
  "error BadSignatureV(uint8 v)",
  "error CannotCancelOrder()",
  "error ConsiderationCriteriaResolverOutOfRange()",
  "error ConsiderationLengthNotEqualToTotalOriginal()",
  "error ConsiderationNotMet(uint256 orderIndex, uint256 considerationIndex, uint256 shortfallAmount)",
  "error CriteriaNotEnabledForItem()",
  "error ERC1155BatchTransferGenericFailure(address token, address from, address to, uint256[] identifiers, uint256[] amounts)",
  "error InexactFraction()",
  "error InsufficientNativeTokensSupplied()",
  "error Invalid1155BatchTransferEncoding()",
  "error InvalidBasicOrderParameterEncoding()",
  "error InvalidCallToConduit(address conduit)",
  "error InvalidConduit(bytes32 conduitKey, address conduit)",
  "error InvalidContractOrder(bytes32 orderHash)",
  "error InvalidERC721TransferAmount(uint256 amount)",
  "error InvalidFulfillmentComponentData()",
  "error InvalidMsgValue(uint256 value)",
  "error InvalidNativeOfferItem()",
  "error InvalidProof()",
  "error InvalidRestrictedOrder(bytes32 orderHash)",
  "error InvalidSignature()",
  "error InvalidSigner()",
  "error InvalidTime(uint256 startTime, uint256 endTime)",
  "error MismatchedFulfillmentOfferAndConsiderationComponents(uint256 fulfillmentIndex)",
  "error MissingFulfillmentComponentOnAggregation(uint8 side)",
  "error MissingItemAmount()",
  "error MissingOriginalConsiderationItems()",
  "error NativeTokenTransferGenericFailure(address account, uint256 amount)",
  "error NoContract(address account)",
  "error NoReentrantCalls()",
  "error NoSpecifiedOrdersAvailable()",
  "error OfferAndConsiderationRequiredOnFulfillment()",
  "error OfferCriteriaResolverOutOfRange()",
  "error OrderAlreadyFilled(bytes32 orderHash)",
  "error OrderCriteriaResolverOutOfRange(uint8 side)",
  "error OrderIsCancelled(bytes32 orderHash)",
  "error OrderPartiallyFilled(bytes32 orderHash)",
  "error PartialFillsNotEnabledForOrder()",
  "error TokenTransferGenericFailure(address token, address from, address to, uint256 identifier, uint256 amount)",
  "error UnresolvedConsiderationCriteria(uint256 orderIndex, uint256 considerationIndex)",
  "error UnresolvedOfferCriteria(uint256 orderIndex, uint256 offerIndex)",
  "error UnusedItemParameters()",
  "event CounterIncremented(uint256 newCounter, address indexed offerer)",
  "event OrderCancelled(bytes32 orderHash, address indexed offerer, address indexed zone)",
  "event OrderFulfilled(bytes32 orderHash, address indexed offerer, address indexed zone, address recipient, (uint8 itemType, address token, uint256 identifier, uint256 amount)[] offer, (uint8 itemType, address token, uint256 identifier, uint256 amount, address recipient)[] consideration)",
  "event OrderValidated(bytes32 orderHash, (address offerer, address zone, (uint8 itemType, address token, uint256 identifierOrCriteria, uint256 startAmount, uint256 endAmount)[] offer, (uint8 itemType, address token, uint256 identifierOrCriteria, uint256 startAmount, uint256 endAmount, address recipient)[] consideration, uint8 orderType, uint256 startTime, uint256 endTime, bytes32 zoneHash, uint256 salt, bytes32 conduitKey, uint256 totalOriginalConsiderationItems) orderParameters)",
  "event OrdersMatched(bytes32[] orderHashes)",
]
`;
